python/salpytools/salpylib.py

```diff
@@ -157,163 +154,220 @@ def reply_to_transition(self, cmdid):
             self.mgr_ackCommand(cmdid, SAL__CMD_COMPLETE, err, message)
 
 
-def validate_transition(current_state, new_state):
-    """
-    Stand-alone function to validate transition. It returns true/false
+class DDSSubscriber(threading.Thread):
+    """Subscribes either to Telemetry or Events and saves the recieved data to 
+    a handed dictionary. In this implementation a new thread is used for every 
+    topic.
+    Attributes:
+        topic: A string of the name of the topic we are subscribing to.
+        handle: An dictionary to save the contents of the recieved data.
+        rate: Rate at which we attempt to obtain new data.
+        timeout: How long to wait for new data before timing out.
     """
-    current_index = csc_states.state_enumeration[current_state]
-    new_index = csc_states.state_enumeration[new_state]
-    transition_is_valid = csc_states.state_matrix[current_index][new_index]
-    if transition_is_valid:
-        LOGGER.info("Transition from {} --> {} is VALID".format(current_state, new_state))
-    else:
-        LOGGER.info("Transition from {} --> {} is INVALID".format(current_state, new_state))
-    return transition_is_valid
+    def __init__(self, topic=None, handle=None, rate=1, timeout=10):
+        threading.Thread.__init__(self)
 
+        self.topic = topic
+        if self.topic is None:
+            raise ValueError("A 'topic' argument must be passed")
 
-class DDSSubscriber(threading.Thread):
+        self.handle = handle
+        if self.handle is None:
+            raise ValueError("A 'handle' argument must be passed")        
 
-    ''' Class to Subscribe to Telemetry, it could a Command (discouraged), Event or Telemetry'''
+        self.rate = rate
+        self.timeout = timeout
 
-    def __init__(self, Device, topic, threadID='1', Stype='Telemetry', tsleep=0.01, timeout=3600, nkeep=100):
-        threading.Thread.__init__(self)
-        self.threadID = threadID
-        self.Device = Device
-        self.topic = topic
+        # Attributes we need to interact with SAL
+        self.subsystem_tag = self.parse_for_subsystem_tag(topic)
+        self.short_topic = self.parse_for_short_topic_name(topic)
+        self.salpy_lib = None
+        self.mgr = None
+        self.data = None
 
-        self.log = create_logger(name=self.Device)
+        # We will turn only a single flag to true on mgr_subscibe_to_topic()
+        self.is_command = False
+        self.is_event = False
+        self.is_telemetry = False
 
-        self.tsleep = tsleep
-        self.Stype  = Stype
-        self.timeout = timeout
-        self.nkeep   = nkeep
-        self.daemon = True
-        self.subscribe()
+    def parse_for_subsystem_tag(self, topic):
```

Comment:
I'm not sure this is worth the trouble. The user can easily pass the topic name and the component so why go through this?

Reply:
I derived this design decision on what I thought was easiest for the user, as well as trying to keep down the amount of arguments needed for readability.

This lead to two different paths.

1) You define the full topic name, nothing more. All other information can be derived from the topic.

2) You define, the topic name, the component, as well as a flag that tells us if it is a command event or topic. 3 arguments as opposed to 1.

Reply:
In understand that... but I think having the 3 arguments is less prone to problems.